com.microsoft.mrtk.core/Utilities/XRSubsystemHelpers.cs

```diff
@@ -51,53 +49,56 @@ public static class XRSubsystemHelpers
         public static void GetAllRunningSubsystemsNonAlloc<T>(List<T> runningSubsystems) where T : ISubsystem
         {
             SubsystemManager.GetSubsystems(runningSubsystems);
             runningSubsystems.RemoveAll(subsystem => !subsystem.running);
         }
 
         /// <summary>
-        /// Get all running subsystems of type T.
+        /// Get all running subsystems of type T. Note, this is internally cached, so
+        /// the list is only valid immediately after invocation.
         /// </summary>
         /// <remarks>
-        /// Caution: this method allocs a new list of subsystems.
-        /// For performance critical (frame loop) applications,
-        /// consider calling SubsystemManager.GetSubsystems directly
-        /// with a pre-allocated list.
+        /// This method allocates on the first invocation with a new, unique type T.
+        /// Subsequent invocations with the same type T will no longer allocate.
         /// </remarks>
         public static List<T> GetAllRunningSubsystems<T>() where T : ISubsystem
         {
-            List<T> filtered = new List<T>();
-
-            foreach (T sys in GetAllSubsystems<T>())
-            {
-                if (sys.running)
-                {
-                    filtered.Add(sys);
-                }
-            }
-
-            return filtered;
+            List<T> subsystems = GetAllSubsystems<T>();
+            subsystems.RemoveAll(subsystem => !subsystem.running);
+            return subsystems;
         }
 
         /// <summary>
-        /// Get all subsystems (not necessarily running) of type T.
+        /// Get all subsystems (not necessarily running) of type T. Note, this is
+        /// internally cached, so the list is only valid immediately after invocation.
         /// </summary>
         /// <remarks>
-        /// Caution: this method allocs a new list of subsystems.
-        /// For performance critical (frame loop) applications,
-        /// consider calling SubsystemManager.GetSubsystems directly
-        /// with a pre-allocated list.
+        /// This method allocates on the first invocation with a new, unique type T.
+        /// Subsequent invocations with the same type T will no longer allocate.
         /// </remarks>
         public static List<T> GetAllSubsystems<T>() where T : ISubsystem
         {
-            List<T> subsystems = new List<T>();
-            SubsystemManager.GetSubsystems(subsystems);
-            return subsystems;
+            // If we haven't queried subsystems of this type before,
+            // we'll need to add a new entry in our cache. This will
+            // allocate the first time, but subsequent queries will
+            // reuse the same cached list. (The list will still be cleared
+            // and re-filled by the GetSubsystems call, but nothing will be
+            // allocated.)
+            if (!scratchCaches.ContainsKey(typeof(T)))
```

Comment:
Based on the performance stats, this caching should be updated post the GetSybsystems call to preserve the List Size and prevent reallocation.  E.G. 

```csharp
            if (!scratchCaches.ContainsKey(typeof(T)))
            {
                scratchCaches.Add(typeof(T), new List<T>());
            }

            List<T> cache = scratchCaches[typeof(T)] as List<T>;
            SubsystemManager.GetSubsystems<T>(cache);
            if(cache.Count > scratchCaches[typeof(T)].Count)
            {
                scratchCaches[typeof(T)] = cache;
            }
```

Or Instead of ```scratchCaches[typeof(T)] = cache;``` use ```scratchCaches[typeof(T)] = new List<T>(cache.Count);```
Which should improve overall allocation if the number of SubSystems is expected to change frequently.

Reply:
Simply put:

* When you do ```new List<T>()``` it creates a list with a capacity of 0
* When the first item is added to the List, in initial capacity is set to 4 (default c# implementation)
* When you add a 5th element, the array inside the list is recreated and the original elements are copied to the new array and the new array becomes the storage of the List.
* Rinse Repeat for adding new items.

So if you only ever expect to have 4 Subsystems (unlikely) caching the new List<T> is ALWAYS going to get a performance hit once the list used by the ```GetSubsystems()``` call (unless internally they are using AddRange)

So while you are saving the "new" cost, you will still be incurring the ListExpansion cost each and every time the ```GetSubsystems()``` call is used.

So the suggestion is to ensure the List that is cached is always a size that is at least as big as the last ```GetSubsystems()``` call and reducing the List Expansion cost.

This is getting into the weeds of performance improvements and may not be needed, but if you are expecting this call to be used every frame (which I wouldn't recommend, as the Subsystems are unlikely to change in the same frame) multiple times, then it would be prudent to be as performant as possible.

Granted, caching the first call per frame and then reusing the same cache in the same frame would also be a good move to reduce the call usage.

If you want to see the evidence of the above (although it is referred to in the MS Docs) simply use the following test:

```csharp
        var myList = new List<int>();
        Debug.Log($"Default capacity of a List: {myList.Capacity}");

        myList.Add(0);
        Debug.Log($"List capacity on use: {myList.Capacity}");
```

Reply:
Yes, the list automatically resizes once you try to add more entries than `Capacity` (I called this out as well, but as a less common case in my post above, since I don't expect the count to change _much_ after startup).

> So while you are saving the "new" cost, you will still be incurring the ListExpansion cost each and every time the GetSubsystems() call is used.

This is incorrect because of the cache this PR is added. This PR is adding a cache so that we only incur the list expansion the _first_ time `GetSubsystems()` is used, per-type. The resulting, pre-expanded list is then retrieved from the dictionary and reused on the next call. The only time it will be expanded again is, as we've both said, if more subsystems are created, Capacity+1 subsystems end up existing at any point, and the list needs to expand. I expect that to be more of an edge case as most subsystems will be created by XR Plugin Management or MRTK3 at startup.

> So the suggestion is to ensure the List that is cached is always a size that is at least as big as the last GetSubsystems() call and reducing the List Expansion cost.

That's exactly what this PR and its cache already does. `List` is a reference type, so the reference that we pass into `SubsystemManager.GetSubsystems` is the exact same as what's in the dictionary. It'll get updated in-place.

Reply:
> That's exactly what this PR and its cache already does.
incorrect, as only new List<T> is cached and passed with a capacity of 0 to the ```GetSubsystems``` call each and every time it is called, and no protection of being called multiple times in the same frame.

All I can do is recommend, it's up to the MS team to manage it.